File: examples/client-example.js
#!/usr/bin/env node
/**
 * JavaScript/Node.js client example for Rust Verusd RPC Server
 *
 * This script demonstrates how to interact with the Verus RPC server
 * using Node.js. It handles authentication, error handling, and provides
 * examples for common RPC methods.
 *
 * Requirements:
 *   npm install node-fetch
 *
 * For Node.js < 18, use:
 *   npm install node-fetch@2
 */

const fetch = require('node-fetch');

/**
 * Simple client for Verus RPC server
 */
class VerusRPCClient {
    /**
     * Initialize the RPC client
     * @param {string} url - The RPC server URL
     * @param {string|null} apiKey - Optional API key for authentication
     * @param {number} timeout - Request timeout in milliseconds
     */
    constructor(url = 'http://localhost:8080', apiKey = null, timeout = 30000) {
        this.url = url;
        this.apiKey = apiKey;
        this.timeout = timeout;
    }

    /**
     * Call an RPC method
     * @param {string} method - The RPC method name
     * @param {Array} params - List of parameters
     * @returns {Promise<any>} The RPC response result
     */
    async call(method, params = []) {
        const headers = {
            'Content-Type': 'application/json',
        };

        // Add API key if provided
        if (this.apiKey) {
            headers['X-API-Key'] = this.apiKey;
        }

        const payload = {
            method,
            params,
        };

        try {
            const controller = new AbortController();
            const timeoutId = setTimeout(() => controller.abort(), this.timeout);

            const response = await fetch(this.url, {
                method: 'POST',
                headers,
                body: JSON.stringify(payload),
                signal: controller.signal,
            });

            clearTimeout(timeoutId);

            if (!response.ok) {
                throw new Error(`HTTP error! status: ${response.status}`);
            }

            const data = await response.json();

            // Check for RPC errors
            if (data.error) {
                throw new Error(
                    `RPC error ${data.error.code}: ${data.error.message}`
                );
            }

            return data.result;
        } catch (error) {
            if (error.name === 'AbortError') {
                throw new Error('Request timeout');
            }
            throw error;
        }
    }

    /**
     * Check server health status
     * @returns {Promise<Object>} Health status
     */
    async healthCheck() {
        const response = await fetch(`${this.url}/health`);
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return response.json();
    }

    /**
     * Check server readiness status
     * @returns {Promise<Object>} Readiness status
     */
    async readinessCheck() {
        const response = await fetch(`${this.url}/ready`);
        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }
        return response.json();
    }

    // Convenience methods for common RPC calls

    /**
     * Get general information about the Verus daemon
     * @returns {Promise<Object>}
     */
    async getInfo() {
        return this.call('getinfo');
    }

    /**
     * Get the current block count
     * @returns {Promise<number>}
     */
    async getBlockCount() {
        return this.call('getblockcount');
    }

    /**
     * Get the block hash at a specific height
     * @param {number} height - Block height
     * @returns {Promise<string>}
     */
    async getBlockHash(height) {
        return this.call('getblockhash', [height]);
    }

    /**
     * Get block data
     * @param {string} blockHash - Block hash
     * @param {number} verbose - Verbosity level (0, 1, or 2)
     * @returns {Promise<Object>}
     */
    async getBlock(blockHash, verbose = 1) {
        return this.call('getblock', [blockHash, verbose]);
    }

    /**
     * Get transaction data
     * @param {string} txid - Transaction ID
     * @param {number} verbose - Verbosity level (0 or 1)
     * @returns {Promise<Object|string>}
     */
    async getTransaction(txid, verbose = 1) {
        return this.call('getrawtransaction', [txid, verbose]);
    }
}

/**
 * Example usage
 */
async function main() {
    // Initialize client (update with your API key if needed)
    const client = new VerusRPCClient(
        'http://localhost:8080',
        'your-secret-api-key-here'  // Remove or update this
    );

    try {
        // Check server health
        console.log('Checking server health...');
        const health = await client.healthCheck();
        console.log('Health:', JSON.stringify(health, null, 2));
        console.log();

        // Get blockchain info
        console.log('Getting blockchain info...');
        const info = await client.getInfo();
        console.log(`Version: ${info.version}`);
        console.log(`Blocks: ${info.blocks}`);
        console.log(`Connections: ${info.connections}`);
        console.log();

        // Get current block count
        console.log('Getting block count...');
        const blockCount = await client.getBlockCount();
        console.log(`Current block count: ${blockCount}`);
        console.log();

        // Get latest block hash
        console.log('Getting latest block hash...');
        const latestHash = await client.getBlockHash(blockCount);
        console.log(`Latest block hash: ${latestHash}`);
        console.log();

        // Get block details
        console.log('Getting block details...');
        const block = await client.getBlock(latestHash);
        console.log(`Block height: ${block.height}`);
        console.log(`Block time: ${block.time}`);
        console.log(`Transactions: ${block.tx ? block.tx.length : 0}`);
        console.log();

    } catch (error) {
        console.error(`Error: ${error.message}`);
        process.exit(1);
    }
}

// Run if called directly
if (require.main === module) {
    main().catch(console.error);
}

module.exports = { VerusRPCClient };

File: examples/client-example.py
#!/usr/bin/env python3
"""
Python client example for Rust Verusd RPC Server

This script demonstrates how to interact with the Verus RPC server
using Python. It handles authentication, error handling, and provides
examples for common RPC methods.

Requirements:
    pip install requests
"""

import requests
import json
from typing import Optional, Dict, Any


class VerusRPCClient:
    """Simple client for Verus RPC server."""

    def __init__(
        self,
        url: str = "http://localhost:8080",
        api_key: Optional[str] = None,
        timeout: int = 30
    ):
        """
        Initialize the RPC client.

        Args:
            url: The RPC server URL
            api_key: Optional API key for authentication
            timeout: Request timeout in seconds
        """
        self.url = url
        self.api_key = api_key
        self.timeout = timeout
        self.session = requests.Session()

        # Set up authentication header if API key is provided
        if api_key:
            self.session.headers.update({
                "X-API-Key": api_key
            })

        # Always set Content-Type
        self.session.headers.update({
            "Content-Type": "application/json"
        })

    def call(self, method: str, params: list = None) -> Dict[str, Any]:
        """
        Call an RPC method.

        Args:
            method: The RPC method name
            params: List of parameters (default: empty list)

        Returns:
            The RPC response result

        Raises:
            requests.exceptions.RequestException: On network errors
            ValueError: On RPC errors
        """
        if params is None:
            params = []

        payload = {
            "method": method,
            "params": params
        }

        try:
            response = self.session.post(
                self.url,
                json=payload,
                timeout=self.timeout
            )
            response.raise_for_status()

            data = response.json()

            # Check for RPC errors
            if "error" in data and data["error"] is not None:
                error = data["error"]
                raise ValueError(
                    f"RPC error {error.get('code')}: {error.get('message')}"
                )

            return data.get("result")

        except requests.exceptions.RequestException as e:
            raise RuntimeError(f"Request failed: {e}")

    def health_check(self) -> Dict[str, str]:
        """Check server health status."""
        response = self.session.get(f"{self.url}/health", timeout=10)
        response.raise_for_status()
        return response.json()

    def readiness_check(self) -> Dict[str, str]:
        """Check server readiness status."""
        response = self.session.get(f"{self.url}/ready", timeout=10)
        response.raise_for_status()
        return response.json()

    # Convenience methods for common RPC calls

    def get_info(self) -> Dict[str, Any]:
        """Get general information about the Verus daemon."""
        return self.call("getinfo")

    def get_block_count(self) -> int:
        """Get the current block count."""
        return self.call("getblockcount")

    def get_block_hash(self, height: int) -> str:
        """Get the block hash at a specific height."""
        return self.call("getblockhash", [height])

    def get_block(self, block_hash: str, verbose: int = 1) -> Dict[str, Any]:
        """Get block data."""
        return self.call("getblock", [block_hash, verbose])

    def get_transaction(self, txid: str, verbose: int = 1) -> Dict[str, Any]:
        """Get transaction data."""
        return self.call("getrawtransaction", [txid, verbose])


def main():
    """Example usage."""
    # Initialize client (update with your API key if needed)
    client = VerusRPCClient(
        url="http://localhost:8080",
        api_key="your-secret-api-key-here"  # Remove or update this
    )

    try:
        # Check server health
        print("Checking server health...")
        health = client.health_check()
        print(f"Health: {json.dumps(health, indent=2)}")
        print()

        # Get blockchain info
        print("Getting blockchain info...")
        info = client.get_info()
        print(f"Version: {info.get('version')}")
        print(f"Blocks: {info.get('blocks')}")
        print(f"Connections: {info.get('connections')}")
        print()

        # Get current block count
        print("Getting block count...")
        block_count = client.get_block_count()
        print(f"Current block count: {block_count}")
        print()

        # Get latest block hash
        print("Getting latest block hash...")
        latest_hash = client.get_block_hash(block_count)
        print(f"Latest block hash: {latest_hash}")
        print()

        # Get block details
        print("Getting block details...")
        block = client.get_block(latest_hash)
        print(f"Block height: {block.get('height')}")
        print(f"Block time: {block.get('time')}")
        print(f"Transactions: {len(block.get('tx', []))}")
        print()

    except Exception as e:
        print(f"Error: {e}")
        return 1

    return 0


if __name__ == "__main__":
    exit(main())
